drivers/sensor/st/lsm6dsv16x/lsm6dsv16x_rtio_stream.c

```diff
@@ -0,0 +1,357 @@
+	fifo_count = (uint16_t)lsm6dsv16x->fifo_status[1] & 0x1U;
+	fifo_count = (fifo_count << 8) | lsm6dsv16x->fifo_status[0];
+	lsm6dsv16x->fifo_count = fifo_count;
+
+	bool has_fifo_ths_trig = fifo_ths_cfg != NULL && fifo_th == 1;
+	bool has_fifo_full_trig = fifo_full_cfg != NULL && fifo_full == 1;
+
+	/* check if no theshold/full fifo interrupt or spurious interrupts */
+	if (!has_fifo_ths_trig && !has_fifo_full_trig) {
+		/* complete operation with no error */
+		rtio_iodev_sqe_ok(sqe->userdata, 0);
+
+		lsm6dsv16x->streaming_sqe = NULL;
+		gpio_pin_interrupt_configure_dt(irq_gpio, GPIO_INT_EDGE_TO_ACTIVE);
+		return;
+	}
+
+	struct rtio_cqe *cqe;
+	int res = 0;
+
+	do {
+		cqe = rtio_cqe_consume(lsm6dsv16x->rtio_ctx);
+		if (cqe != NULL) {
+			if ((cqe->result < 0) && (res == 0)) {
+				LOG_ERR("Bus error: %d", cqe->result);
+				res = cqe->result;
+			}
+			rtio_cqe_release(lsm6dsv16x->rtio_ctx, cqe);
+		}
+	} while (cqe != NULL);
+
+	/* Bail/cancel attempt to read sensor on any error */
+	if (res != 0) {
+		rtio_iodev_sqe_err(lsm6dsv16x->streaming_sqe, res);
+		lsm6dsv16x->streaming_sqe = NULL;
+		return;
+	}
+
+	enum sensor_stream_data_opt data_opt;
+
+	if (has_fifo_ths_trig && !has_fifo_full_trig) {
+		/* Only care about fifo threshold */
+		data_opt = fifo_ths_cfg->opt;
+	} else if (!has_fifo_ths_trig && has_fifo_full_trig) {
+		/* Only care about fifo full */
+		data_opt = fifo_full_cfg->opt;
+	} else {
+		/* Both fifo threshold and full */
+		data_opt = MIN(fifo_ths_cfg->opt, fifo_full_cfg->opt);
+	}
+
+	if (data_opt == SENSOR_STREAM_DATA_NOP || data_opt == SENSOR_STREAM_DATA_DROP) {
+		uint8_t *buf;
+		uint32_t buf_len;
+
+		/* Clear streaming_sqe since we're done with the call */
+		if (rtio_sqe_rx_buf(lsm6dsv16x->streaming_sqe, sizeof(struct lsm6dsv16x_fifo_data),
+				    sizeof(struct lsm6dsv16x_fifo_data), &buf, &buf_len) != 0) {
+			rtio_iodev_sqe_err(lsm6dsv16x->streaming_sqe, -ENOMEM);
+			lsm6dsv16x->streaming_sqe = NULL;
+			gpio_pin_interrupt_configure_dt(irq_gpio, GPIO_INT_EDGE_TO_ACTIVE);
+			return;
+		}
+
+		struct lsm6dsv16x_fifo_data *rx_data = (struct lsm6dsv16x_fifo_data *)buf;
+
+		memset(buf, 0, buf_len);
+		rx_data->header.is_fifo = 1;
+		rx_data->header.timestamp = k_ticks_to_ns_floor64(k_uptime_ticks());
+		rx_data->int_status = lsm6dsv16x->fifo_status[1];
+		rx_data->fifo_count = 0;
+
+		/* complete request with ok */
+		rtio_iodev_sqe_ok(lsm6dsv16x->streaming_sqe, 0);
+		lsm6dsv16x->streaming_sqe = NULL;
+
+		if (data_opt == SENSOR_STREAM_DATA_DROP) {
+
+			/*
+			 * Flush the FIFO by setting the mode to LSM6DSV16X_BYPASS_MODE
+			 *
+			 * STMEMSC API equivalent code:
+			 *
+			 *   lsm6dsv16x_fifo_mode_set(ctx, LSM6DSV16X_BYPASS_MODE);
+			 */
+			struct rtio_sqe *write_fifo_mode = rtio_sqe_acquire(lsm6dsv16x->rtio_ctx);
+			uint8_t lsm6dsv16x_fifo_mode_set[] = {
+				LSM6DSV16X_FIFO_CTRL4,
+				LSM6DSV16X_BYPASS_MODE,
+			};
+
+			rtio_sqe_prep_tiny_write(write_fifo_mode, iodev,
```

Comment:
Since you don't want a response here presumably, probably safer to ensure no completion is generated...

```c
write_fifo_mode->flags |= RTIO_SQE_NO_RESPONSE;
```

It's also possible that this is still in flight when the next gpio interrupt comes as you unmask interrupts below. This might lead to the kind of thing @XenuIsWatching saw https://github.com/zephyrproject-rtos/zephyr/pull/81447/files#r1852758137

Reply:
rtio_sqe_prep_callback_no_cqe() is actually a better answer as it sets the flag in the prep helper